### ISSUE TITLE
Different privileges in containers using the same network namespace

### ISSUE THREAD

**Issue by miheinr** [containers/podman#19185]
### Issue Description

As discussed with @vrothberg we have an use case where we run containers within the same network namespace.  For that we run the first container (e.g. app1) in the background and "connect" to the network namespace of app1 by starting the second container (app2) with the podman-run option `--network container:app1`.  In our use case we flush and apply netfilter rules in the (shared) network namespace by using e.g. `nft flush ruleset` in app2. From `man capabilities` we know that the capability CAP_NET_ADMIN is necessary for that.
We discovered that there is a difference in priviliges in app2 if we run app1 with `--userns=keep-id` or not:

- If app1 runs **without** `--userns=keep-id` then we are not able to successfully run `nft flush ruleset` in app2 (which is expected) because `podman top app2 capeff` shows that CAP_NET_ADMIN is missing. If we use `--cap-add CAP_NET_ADMIN` for app2 then `nft flush ruleset` runs without error.
- whereas if we run app1 **with** `--userns=keep-id` then we are able to successfully run `nft flush ruleset` in app2 although `podman top app2 capeff` shows that CAP_NET_ADMIN is missing.  **This is not expected.**

### Steps to reproduce the issue

Steps to reproduce the issue
1. Build an image (`podman build -t ubi8nft .`) in the same directory as following `Containerfile`
`FROM ubi8`
`RUN dnf -y install nftables`
2. run app1: 
`podman run -dt --name app1 ubi8nft:latest /bin/bash`
3. run app2 with the same network: 
`podman run -dt --network container:app1 --name app2 ubi8nft:latest /bin/bash`
4. run `nft flush ruleset` in app2 where **it is expected to get an error:**
`podman exec -it app2 nft flush ruleset`
The error should be:
`Error: Could not process rule: Operation not permitted`
`flush ruleset`
`^^^^^^^^^^^^^^`
5. Check capabilities with
`podman top app2 capeff`
Output should be (without CAP_NET_ADMIN):
`EFFECTIVE CAPS`
`CHOWN,DAC_OVERRIDE,FOWNER,FSETID,KILL,NET_BIND_SERVICE,NET_RAW,SETFCAP,SETGID,SETPCAP,SETUID,SYS_CHROOT`
6. remove app1 and app2 containers:
`podman rm -f app2`
`podman rm -f app1`
7. the same procedure where we run app1 with option `--userns=keep-id`:
`podman run -dt --userns=keep-id --name app1 ubi8nft:latest /bin/bash`
`podman run -dt --network container:app1 --name app2 ubi8nft:latest /bin/bash`
8. This command runs without error which **is not expected**:
`podman exec -it app2 nft flush ruleset`
9. Check capabilities with
`podman top app2 capeff`
Output should be (without CAP_NET_ADMIN):
`EFFECTIVE CAPS`
`CHOWN,DAC_OVERRIDE,FOWNER,FSETID,KILL,NET_BIND_SERVICE,NET_RAW,SETFCAP,SETGID,SETPCAP,SETUID,SYS_CHROOT`

### Describe the results you received

Although we have the same effective capabilities in both cases (5. and 9.) we see different privileges in containers app2.

### Describe the results you expected

We would expect to see the same error message in 8. as in 4.

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.29.0
  cgroupControllers: []
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.1.6-1.module+el8.8.0+18098+9b44df5f.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.6, commit: 8c4ab5a095127ecc96ef8a9c885e0e1b14aeb11b'
  cpuUtilization:
    idlePercent: 99.57
    systemPercent: 0.12
    userPercent: 0.31
  cpus: 2
  distribution:
    distribution: '"rhel"'
    version: "8.8"
  eventLogger: file
  hostname: rhel8
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 4.18.0-477.15.1.el8_8.x86_64
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 317919232
  memTotal: 3811012608
  networkBackend: cni
  ociRuntime:
    name: runc
    package: runc-1.1.4-1.module+el8.8.0+18060+3f21f2cc.x86_64
    path: /usr/bin/runc
    version: |-
      runc version 1.1.4
      spec: 1.0.2-dev
      go: go1.19.4
      libseccomp: 2.5.2
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_SYS_CHROOT,CAP_NET_RAW,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-2.module+el8.8.0+18060+3f21f2cc.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.2
  swapFree: 2145382400
  swapTotal: 2147479552
  uptime: 9h 8m 39.00s (Approximately 0.38 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.access.redhat.com
  - registry.redhat.io
  - docker.io
store:
  configFile: /home/a450665/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 2
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/a450665/.local/share/containers/storage
  graphRootAllocated: 19001245696
  graphRootUsed: 10810925056
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 12
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/a450665/.local/share/containers/storage/volumes
version:
  APIVersion: 4.4.1
  Built: 1686839996
  BuiltTime: Thu Jun 15 16:39:56 2023
  GitCommit: ""
  GoVersion: go1.19.6
  Os: linux
  OsArch: linux/amd64
  Version: 4.4.1
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

No

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

**Comment by rhatdan** [comment]
Does
```
 podman exec --user root -it app2 nft flush ruleset
```
Work? You have to be root to run nft, and in the case you run with --userns=keep-id the default user is the user running podman's UID, not root.

**Comment by vrothberg** [comment]
Thanks for opening the issue, @miheinr !

@rhatdan, the confusing is part is step 8)

> This command runs without error which is not expected:

Container `app2` is _not_ started with `--userns=keep-id` but joins the nw NS of `app1` which was created with `--userns=keep-id`.

**Comment by Luap99** [comment]
I am pretty sure this is normal kernel behavior, mixing user namespaces with other namespaces is a bad idea in general.

From user_namespaces(7):
>When a user namespace is created, the kernel records the
          effective user ID of the creating process as being the "owner"
          of the namespace.  A process that resides in the parent of the
          user namespace and whose effective user ID matches the owner
          of the namespace has all capabilities in the namespace.

app2 is in the parent user namespace so it has all caps for the child userns.

also read `Interaction of user namespaces and other types of namespaces`
> When a new namespace (other than a user namespace) is created via
       [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html) or [unshare(2)](https://man7.org/linux/man-pages/man2/unshare.2.html), the kernel records the user namespace of
       the creating process as the owner of the new namespace.  (This
       association can't be changed.)  When a process in the new
       namespace subsequently performs privileged operations that
       operate on global resources isolated by the namespace, the
       permission checks are performed according to the process's
       capabilities in the user namespace that the kernel associated
       with the new namespace.

So basically app2 is part of the parent userns while app1 has the netns created in the child userns. Thus app2 as parent always has all caps for all namespaces created by app1, the unset caps for app2 are only effective in its own (parent) userns all namespaces associated with that.


I don't think there is any way to prevent that other than making sure that app2 is not in the parent userns from app1.
So for app2 you can either add `--userns container:app1` (use the same usersns) or `--userns keep-id` (use a different userns) then it should fail correctly.

@giuseppe WDYT?

**Comment by giuseppe** [comment]
Yes that is correct. app2 is running from the parent user namespace so it has all the capabilities, you need to enforce the same userns if you want to keep the same behavior

**Comment by Luap99** [comment]
Should we throw a warning in podman when this is done? This seems like a dangerous footgun otherwise as a a user I would totally expect the container to honour the given capabilities.

**Comment by giuseppe** [comment]
Should we print a warning each time we use a shared namespace?  Wouldn't that be too noisy if users are expecting that?

**Comment by vrothberg** [comment]
> Wouldn't that be too noisy if users are expecting that?

Hard to predict.  I think this depends on the use case.  It doesn't make sense in a Pod but the behavior was quite surprising.  A first step could be to add a warning to all options in the man pages that would join a shared namespace.

**Comment by Luap99** [comment]
Yeah it could get noisy if the user expects that. Making this clear in docs is definitely a good start.

>  It doesn't make sense in a Pod

For pods we enforce that the userns is set on the pod and all container will join the userns so it should not be a problem there.

For a warning we should really only print it when we know the the container is in a parent ns of the target container. 

---

Another idea would be to create a private userns for each container by default while keeping the same uid mapping. Then we can be sure to never be the parent of another container. This of course would be a larger, likely breaking change but I think it worth to throw the idea out.

**Comment by giuseppe** [comment]
A private userns would be a breaking change since then you would not be able to access resources created in the `podman unshare` environment.

**Comment by miheinr** [comment]
Thank you all for the explanation and discussion.  I think in our use case where we modify the firewall rules of app1 in app2 we will use "--network container:app1 --userns container:app1 --cap-add CAP_NET_ADMIN" to avoid any confusion.

**Comment by rhatdan** [comment]
Since this is not something we want to support, and you have a workaround closing.

**Comment by Luap99** [comment]
We still should point this out in the documentation.

**Comment by Luap99** [comment]
FYI I wrote a blog about this a while back https://blog.podman.io/2023/12/interaction-between-user-namespaces-and-capabilities/

I was thinking about adding this to man page somewhere but couldn't find a good place  as it is quite complex to describe and would apply to every namespace option
